core/src/main/java/com/google/errorprone/bugpatterns/EmptyCatch.java

```diff
@@ -0,0 +1,372 @@
+import static com.sun.source.tree.Tree.Kind.ENHANCED_FOR_LOOP;
+import static com.sun.source.tree.Tree.Kind.SWITCH;
+import static com.sun.source.tree.Tree.Kind.TRY;
+import static com.sun.source.tree.Tree.Kind.VARIABLE;
+
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.CatchTree;
+import com.sun.source.tree.StatementTree;
+import com.sun.source.tree.TryTree;
+import com.sun.source.tree.BlockTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.ExpressionStatementTree;
+
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+
+import java.util.List;
+
+/**
+ * A bug checker for empty exception handling block:
+ *
+ * <pre>
+ * {@code
+ * try {
+ *   // do something
+ * } catch (XX e) {
+ *   // empty, only comment, or only log printing statements
+ * }
+ * }
+ * </pre>
+ * We also implement a variety of false positive suppression heuristics.
+ * <p>
+ * For more detail, refer to the paper:
+ * "Simple Testing Can Prevent Most Critical Failures: 
+ *  An Analysis of Production Failures in Distributed Data-intensive Systems"
+ *  Yuan et al. Proceedings of the 11th USENIX Symposium on Operating Systems Design 
+ *  and Implementation (OSDI), 2014
+ *
+ * @author yuan@eecg.utoronto.ca (Ding Yuan)
+ */
+@BugPattern(name = "EmptyCatch",
+    summary = "Exception handling block is empty",
+    explanation = "It is rarely correct to do nothing in response to a caught exception. In production systems, "
+            + "Murphy's law often applies: Anything that can go wrong, will go wrong. "
+            + "Therefore it is particularly important to properly handle the exceptions "
+            + "regardless how rarely they are likely to occur.\n\n"
+            + "Read Effective Java, Item 65, \"Don't ignore exceptions\"; "
+            + "[Google Java Style guide section 6.2] "
+            + "(https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions); and "
+            + "\"[Simple Testing Can Prevent Most Critical Failures] "
+            + "(http://www.eecg.toronto.edu/~yuan/papers/failure_analysis_osdi14.pdf)\" "
+            + "for more detailed discussions on the harm of swallowing the exceptions. ",
+    category = JDK, maturity = EXPERIMENTAL, severity = WARNING)
+public class EmptyCatch extends BugChecker implements TryTreeMatcher {
+  /* If one of the following exceptions is ignored, we suppress the warning. */
+  ImmutableSet<String> harmlessExceptions = ImmutableSet.of("java.io.FileNotFoundException", 
+    "java.lang.ReflectiveOperationException", "java.lang.ClassNotFoundException",
+    "java.lang.reflect.InvocationTargetException", "java.lang.NoSuchFieldException",
+    "java.lang.NoSuchMethodException", "java.lang.InstantiationException");
+  
+  /* If the method is one of the following, we allow it to ignore exceptions. */
+  String[] whitelistedMethods = {
```

Comment:
These methods are related with system shutdown or resource disposal. We observe it's quite common that exceptions from these methods are ignored and it is unclear whether anything meaningful can be done to handle these exceptions.